src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Serialization/MessagePack/Formatters/RazorProjectInfoFormatter.cs

```diff
@@ -18,37 +18,43 @@ private RazorProjectInfoFormatter()
 
     public override RazorProjectInfo Deserialize(ref MessagePackReader reader, SerializerCachingOptions options)
     {
-        reader.ReadArrayHeaderAndVerify(7);
+        VerifyVersionNumber(ref reader);
 
-        var version = reader.ReadInt32();
+        reader.ReadArrayHeaderAndVerify(4);
```

Comment:
Why not have this above the verify, and just have the reader move in `VerifyVersionNumber` since it's already passed by ref?

Reply:
I guess because that felt a bit more awkward to me when I wrote this. The problem is that we need to verify the array header length after the version check. Otherwise, we may never get to the version check before we blow up because the array size changed if a new version happens to add or remove `RazorProjectInfo` field. I felt that it was better to have a separate method to verify the version by peeking ahead and then leaving this method to do the real data read. 

> since it's already passed by ref?

`MessagePackReader` really should be passed by ref in some way. It's a ref struct and a ref has already been taken for it. I'll update `VerifyVersionNumber` to `ref readonly`.